p2p/message.go

```diff
@@ -323,6 +331,54 @@ func (bm *BlockResponseMessage) Decode(r io.Reader) error {
 	return nil
 }
 
+type TransactionMessage struct {
+	Extrinsics []common.Extrinsic
+}
+
+func (tm *TransactionMessage) String() string {
+	return fmt.Sprintf("TransactionMessage extrinsics=0x%x", tm.Extrinsics)
+}
+
+func (tm *TransactionMessage) Encode() ([]byte, error) {
+	// scale encode each extrinsic
+	var encodedExtrinsics = make([]byte, 0)
+	for _, extrinsic := range tm.Extrinsics {
+		encExt, err := scale.Encode([]byte(extrinsic))
+		if err != nil {
+			return nil, err
+		}
+		encodedExtrinsics = append(encodedExtrinsics, encExt...)
+	}
+
+	// scale encode the set of all extrinsics
+	encodedMessage, err := scale.Encode(encodedExtrinsics)
+
+	// prepend message type to message
+	return append([]byte{TransactionMsgType}, encodedMessage...), err
+}
+
+//Decodes the message into a TransactionMessage, it assumes the type byte han been removed
+func (tm *TransactionMessage) Decode(r io.Reader) error {
+	sd := scale.Decoder{Reader: r}
+	decodedMessage, err := sd.Decode([]byte{})
+	if err != nil {
+		return err
+	}
+	messageSize := len(decodedMessage.([]byte))
+	bytesProcessed := 0
+	// loop through the message decoding extrinsics until they have all been decoded
+	for bytesProcessed < messageSize {
+		decodedExtrinsic, err := scale.Decode(decodedMessage.([]byte)[bytesProcessed:], []byte{})
```

Comment:
we could do this using a scale.Decoder{reader} so that it reads the next byte array (extrinsic) then advances the buffer then you keep looping until the buffer is empty
eg. something like:
```
sd := scale.Decoder{Reader: r}
for sd.Reader not empty {
     decodedExtrinsic := sd.Decode([]byte{})
     tm.Extrinsics = append(tm.Extrinsics, decodedExtrinsic)
}
```
this should work since each call to `sd.Decode` only reads up to the length of the extrinsic

Reply:
I'm a little confused, I am using scale.Decoder in line 362, then in 363 I decode the message.  After this point the reader is at EOF, so in the loop sd.Reader is empty so it won't do further decoding.  The message has to be decoded twice, I'm doing the second encoding starting at 369.  I there a way I can do this with a reader?  Do I need to init a second reader for this?

Reply:
yeah I was thinking you make another reader that contains the first decoded message, then continue to read each encoded extrinsic through the loop. this way you don't need to keep track of bytesProcessed. however I'm not sure how much practical difference this makes so I'm fine with this